File: internal/agent/registry.go
package agent

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"time"
)

// handler is the type-erased form of a registered action.
type handler struct {
	version int
	// timeout overrides the server's default budget. Zero means use it.
	timeout time.Duration
	fn      func(context.Context, json.RawMessage) (any, error)
}

// Registry holds the actions an agent will serve. Actions are registered at
// startup and never change, so lookups need no lock beyond construction.
type Registry struct {
	mu      sync.RWMutex
	actions map[string]handler
}

// NewRegistry returns an empty registry.
func NewRegistry() *Registry {
	return &Registry{actions: make(map[string]handler)}
}

// Register adds a typed action. In and Out are ordinary structs; the registry
// handles JSON decoding and encoding, so a handler never sees raw bytes and
// cannot forget to validate the shape of its input.
//
// It panics on a duplicate name: that is a programming error discovered at
// startup, not a runtime condition.
func Register[In, Out any](r *Registry, name string, version int, fn func(context.Context, In) (Out, error)) {
	register(r, name, version, 0, fn)
}

// RegisterSlow is Register for an action that legitimately takes longer than
// the server's default budget -- archiving a customer's whole home directory,
// restoring one, dumping a large database. Without it the only ways to allow
// those are to raise the budget for every action, which lets a wedged handler
// hold a slot for an hour, or to build a job queue, which is a lot of
// machinery for a handful of operations.
func RegisterSlow[In, Out any](r *Registry, name string, version int, timeout time.Duration,
	fn func(context.Context, In) (Out, error)) {
	register(r, name, version, timeout, fn)
}

func register[In, Out any](r *Registry, name string, version int, timeout time.Duration,
	fn func(context.Context, In) (Out, error)) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if _, dup := r.actions[name]; dup {
		panic("agent: action registered twice: " + name)
	}
	r.actions[name] = handler{
		version: version,
		timeout: timeout,
		fn: func(ctx context.Context, raw json.RawMessage) (any, error) {
			var in In
			if len(raw) > 0 {
				dec := json.NewDecoder(bytes.NewReader(raw))
				dec.DisallowUnknownFields() // a typo in a field name must fail, not be ignored
				if err := dec.Decode(&in); err != nil {
					return nil, &PayloadError{Err: err}
				}
			}
			if v, ok := any(&in).(Validator); ok {
				if err := v.Validate(); err != nil {
					return nil, &PayloadError{Err: err}
				}
			}
			return fn(ctx, in)
		},
	}
}

// Validator lets an input struct check itself. Implement it on the pointer
// receiver; the registry calls it after decoding and before the handler runs.
type Validator interface {
	Validate() error
}

// PayloadError marks a caller-side input problem so the server can answer
// with CodeBadPayload rather than CodeInternal.
type PayloadError struct{ Err error }

func (e *PayloadError) Error() string { return "invalid payload: " + e.Err.Error() }
func (e *PayloadError) Unwrap() error { return e.Err }

// DeniedError marks a refusal by policy rather than a failure.
type DeniedError struct{ Reason string }

func (e *DeniedError) Error() string { return "denied: " + e.Reason }

// Budget returns how long the named action may run, falling back to def for
// an action that did not ask for more (and for one that does not exist -- the
// lookup that follows reports that properly).
func (r *Registry) Budget(name string, def time.Duration) time.Duration {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if h, ok := r.actions[name]; ok && h.timeout > 0 {
		return h.timeout
	}
	return def
}

// lookup resolves an action and checks the requested version.
func (r *Registry) lookup(name string, version int) (handler, error) {
	r.mu.RLock()
	h, ok := r.actions[name]
	r.mu.RUnlock()
	if !ok {
		return handler{}, ErrUnknownAction{Name: name}
	}
	// Zero means "whatever the agent has", for opanelctl and for anybody
	// holding a socket open by hand. Every call from the panel pins a
	// version, which is where the guarantee is needed and where it applies.
	if version != 0 && version != h.version {
		return handler{}, ErrVersionMismatch{Name: name, Has: h.version, Wanted: version}
	}
	return h, nil
}

// ErrUnknownAction and ErrVersionMismatch are distinct types because the
// caller has to tell them apart to pick a code, and it used to guess: it
// reported version_mismatch whenever the request carried a version, which
// every real request does. A mistyped action name therefore came back as a
// version problem, which is the one answer guaranteed to send somebody
// looking in the wrong place.
type ErrUnknownAction struct{ Name string }

func (e ErrUnknownAction) Error() string { return fmt.Sprintf("unknown action %q", e.Name) }

// ErrVersionMismatch means the action exists but not at the version asked for.
type ErrVersionMismatch struct {
	Name   string
	Has    int
	Wanted int
}

func (e ErrVersionMismatch) Error() string {
	return fmt.Sprintf("action %q is version %d, caller asked for %d", e.Name, e.Has, e.Wanted)
}

// Actions lists registered action names with their versions, sorted. Used by
// opanelctl to show what an agent build supports.
func (r *Registry) Actions() map[string]int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make(map[string]int, len(r.actions))
	for k, v := range r.actions {
		out[k] = v.version
	}
	return out
}

File: crates/opanel-agent/src/registry.rs
//! The set of things the agent will do, and nothing else.
//!
//! An action is looked up by name and version. There is no path from a
//! request to a shell, and no way to add an action at runtime: the table is
//! built at startup and then only read.

use crate::protocol::{Code, Request, Response};
use serde_json::Value;
use std::collections::BTreeMap;
use std::time::Duration;

/// What an action returns: a JSON payload, or a coded failure.
pub type Outcome = Result<Value, Failure>;

/// A failure with the code the caller will branch on.
#[derive(Debug)]
pub struct Failure {
    pub code: Code,
    pub message: String,
}

impl Failure {
    pub fn new(code: Code, message: impl Into<String>) -> Self {
        Self { code, message: message.into() }
    }

    pub fn internal(message: impl Into<String>) -> Self {
        Self::new(Code::Internal, message)
    }

    pub fn bad_payload(message: impl Into<String>) -> Self {
        Self::new(Code::BadPayload, message)
    }
}

impl From<std::io::Error> for Failure {
    fn from(e: std::io::Error) -> Self {
        Failure::internal(e.to_string())
    }
}

/// One registered operation.
pub struct Action {
    pub version: i64,
    /// Overrides the server's default budget. Zero means use it.
    pub timeout: Duration,
    pub run: Box<dyn Fn(Option<Value>) -> Outcome + Send + Sync>,
}

/// The table.
#[derive(Default)]
pub struct Registry {
    actions: BTreeMap<String, Action>,
}

impl Registry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Registers an action. Panics on a duplicate name: two handlers for one
    /// name is a build mistake, and a root daemon should not start with one.
    pub fn register<F>(&mut self, name: &str, version: i64, run: F)
    where
        F: Fn(Option<Value>) -> Outcome + Send + Sync + 'static,
    {
        self.register_slow(name, version, Duration::ZERO, run);
    }

    /// `register` for an action that legitimately takes longer than the
    /// default budget -- archiving a home directory, dumping a database.
    pub fn register_slow<F>(&mut self, name: &str, version: i64, timeout: Duration, run: F)
    where
        F: Fn(Option<Value>) -> Outcome + Send + Sync + 'static,
    {
        let existing = self
            .actions
            .insert(name.to_owned(), Action { version, timeout, run: Box::new(run) });
        assert!(existing.is_none(), "action registered twice: {name}");
    }

    /// What this build serves, so the panel can detect a version skew rather
    /// than guessing from its own compiled-in list.
    pub fn names(&self) -> BTreeMap<String, i64> {
        self.actions.iter().map(|(k, a)| (k.clone(), a.version)).collect()
    }

    pub fn len(&self) -> usize {
        self.actions.len()
    }

    /// The budget this action gets, falling back to the server's default.
    pub fn budget(&self, name: &str, default: Duration) -> Duration {
        match self.actions.get(name) {
            Some(a) if !a.timeout.is_zero() => a.timeout,
            _ => default,
        }
    }

    /// Looks the action up and runs it.
    pub fn dispatch(&self, req: &Request) -> Response {
        let Some(action) = self.actions.get(&req.action) else {
            return Response::failed(
                &req.id,
                Code::UnknownAction,
                format!("unknown action {:?}", req.action),
            );
        };
        // Zero means "whatever the agent has", matching the Go agent: it is
        // for opanelctl and for anybody holding the socket open by hand.
        // Every call from the panel pins a version, which is where the
        // guarantee is needed and where it applies.
        if req.version != 0 && action.version != req.version {
            return Response::failed(
                &req.id,
                Code::VersionMismatch,
                format!(
                    "action {:?} is version {} here, caller asked for {}",
                    req.action, action.version, req.version
                ),
            );
        }
        match (action.run)(req.payload.clone()) {
            Ok(payload) => Response::ok(&req.id, payload),
            Err(f) => Response::failed(&req.id, f.code, f.message),
        }
    }
}
